src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java

```diff
@@ -185,144 +179,234 @@ public void startJam() {
-			if (inOvertime) {
-				setInOvertime(false);
-			}
+			_startTimeout();
 			requestBatchEnd();
+
+			ScoreBoardManager.gameSnapshot();
 		}
 	}
-	private void _stopTimeout() {
-		synchronized (runLock) {
-			createSnapshot(ACTION_STOP_TO);
-			
-			ClockModel tc = getClockModel(Clock.ID_TIMEOUT);
-			ClockModel lc = getClockModel(Clock.ID_LINEUP);
-			ClockModel pc = getClockModel(Clock.ID_PERIOD);
-			ClockModel ic = getClockModel(Clock.ID_INTERMISSION);
-			
+	public void setTimeoutType(String owner, boolean review) {
+		ClockModel tc = getClockModel(Clock.ID_TIMEOUT);
+
+		requestBatchStart();
+		if (!tc.isRunning()) {
+			timeout();
+		}
+		setTimeoutOwner(owner);
+		setOfficialReview(review);
+		if (owner != "" && owner != "O") {
+			restartPcAfterTimeout = false;
+		}
+		requestBatchEnd();
+	}
+	private void _preparePeriod() {
+		ClockModel pc = getClockModel(Clock.ID_PERIOD);
+		ClockModel jc = getClockModel(Clock.ID_JAM);
+		ClockModel ic = getClockModel(Clock.ID_INTERMISSION);
+
+		requestBatchStart();
+		pc.setNumber(ic.getNumber()+1);
+		pc.resetTime();
+		restartPcAfterTimeout = false;
+		if (settings.getBoolean("ScoreBoard." + Clock.ID_JAM + ".ResetNumberEachPeriod")) {
+			jc.setNumber(jc.getMinimumNumber());
+		}
+		for (TeamModel t : getTeamModels()) {
+			t.resetTimeouts(false);
+		}		
+		requestBatchEnd();
+	}
+	private void _possiblyEndPeriod() {
+		ClockModel pc = getClockModel(Clock.ID_PERIOD);
+		ClockModel jc = getClockModel(Clock.ID_JAM);
+		ClockModel tc = getClockModel(Clock.ID_TIMEOUT);
+
+		if (pc.isTimeAtEnd() && !pc.isRunning() && !jc.isRunning() && !tc.isRunning()) {
 			requestBatchStart();
-			if (pc.isTimeAtEnd()) {
-				ic.resetTime(); //TODO: add a rule to make this optional
-				ic.start();
-			} else {
-				lc.resetTime();
-				lc.start();
-			}
-			tc.stop();
+			setInPeriod(false);
+			setOfficialScore(false);
+			_endLineup();
+			_startIntermission();
 			requestBatchEnd();
 		}
 	}
-	private void _startLineup() {
-		synchronized (runLock) {
-			createSnapshot(ACTION_LINEUP);
-			
-			ClockModel lc = getClockModel(Clock.ID_LINEUP);
-			ClockModel ic = getClockModel(Clock.ID_INTERMISSION);
+	private void _startJam() {
+		ClockModel pc = getClockModel(Clock.ID_PERIOD);
+		ClockModel jc = getClockModel(Clock.ID_JAM);
+
+		requestBatchStart();
+		_endIntermission(false);
+		_endTimeout();
+		_endLineup();
+		setInPeriod(true);
+		pc.start();
+		jc.startNext();
+
+		getTeamModel("1").startJam();
+		getTeamModel("2").startJam();
+		requestBatchEnd();
+	}
+	private void _endJam(boolean force) {
+		ClockModel pc = getClockModel(Clock.ID_PERIOD);
+		ClockModel jc = getClockModel(Clock.ID_JAM);
 
-			ic.stop();
+		if (!jc.isRunning() && !force) { return; }
+		
+		requestBatchStart();
+		jc.stop();
+		getTeamModel("1").stopJam();
+		getTeamModel("2").stopJam();
+		setInOvertime(false);
 
-			requestBatchStart();
-			lc.resetTime();
-			lc.start();
-			requestBatchEnd();
+		if (pc.getTimeRemaining() < 30000) {
```

Comment:
Should we use the normal lineup time here rather than hardcoding?

Reply:
In the WFTDA rules it is hardcoded to 30s, not duration of lineup.
And the proper way would be to make it configurable in the ruleset. I'll add a TODO note for when I add the other items I have on my list for the rulesets.

Reply:
The duration of a lineup can be set with the "Clock.Lineup.Time" rule.